tests/src/test/scala/cats/tests/TraverseTests.scala

```diff
@@ -0,0 +1,53 @@
+package cats
+package tests
+
+import org.scalatest.prop.PropertyChecks
+import org.scalacheck.Arbitrary
+
+import cats.instances.all._
+
+abstract class TraverseCheck[F[_]: Traverse](name: String)(implicit ArbFInt: Arbitrary[F[Int]]) extends CatsSuite with PropertyChecks {
+
+  test(s"Traverse[$name].indexed") {
+    forAll { (fa: F[Int]) =>
+      fa.indexed.toList should === (fa.toList.zipWithIndex)
+    }
+  }
+
+  test(s"Traverse[$name].mapWithIndex") {
+    forAll { (fa: F[Int]) =>
+      fa.mapWithIndex((a, i) => (a, i)).toList should === (fa.toList.zipWithIndex)
+    }
+  }
+
+  test(s"Traverse[$name].traverseWithIndex") {
+    forAll { (fa: F[Int]) =>
+      fa.traverseWithIndex((a, i) => Option((a, i))).map(_.toList) should === (Option(fa.toList.zipWithIndex))
```

Comment:
could we have a stronger test, like take `fn: ((Int, Int)) => Option[Int]` and then
```scala
val left = fa.traverseWithIndex((a, i) => fn((a, i))).map(_.toList)
val right = (fa.toList.zipWithIndex match {
  case Nil => Some(Nil)
  case h :: tail => tail.foldM(h :: Nil) { case (l, ai) => fn(ai).map(_ :: l) }.map(_.reverse)
})
left should === right
```